<a href="https://colab.research.google.com/github/ChristianHopf/B455/blob/main/B455_project2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this implementation, I used TensorFlow's Keras API.
To determine the hyper-parameters, I first studied the Keras documentation.

For the baseline model, I decided to use:
- 4 hidden layers
- 128 neurons per layer
- 10 output neurons (for the 10 possible classes)
- ReLU activation
- Adam optimizer
- Categorical Cross Entropy loss function

In [ ]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

Training and testing data

x_train: training data without labels

y_train: training data labels

x_test: testing data without labels

y_test: testing data with labels

In [43]:
(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar10.load_data()

Before building the model, I check the shapes of the training and testing data. I also scale down pixel values of x_train and x_test from 0 to 255 to be between 0 and 1

In [44]:
print("x_train shape: ", np.shape(x_train))
print("y_train shape: ", np.shape(y_train))
print("x_test shape: ", np.shape(x_test))
print("y_test shape: ", np.shape(y_test))

x_train = x_train / 255.0
x_test = x_test / 255.0

x_train shape:  (50000, 32, 32, 3)
y_train shape:  (50000, 1)
x_test shape:  (10000, 32, 32, 3)
y_test shape:  (10000, 1)


Baseline model: TensorFlow Sequential model with 4 hidden layers, using ReLU activation.

Compiled with Adam optimizer (default parameters), because it is efficient when training models on large datsets, and the SparseCategoricalCrossentropy loss function, because there are more than 2 output labels. I first tried to use CategoricalCrossentropy, but when training the model, I encountered a ValueError: Shapes (None, 1) and (None, 10) are incompatible. On reading the documentation for CategoricalCrossentropy loss, I found it expected the labels to be in one-hot representation. I changed my loss function to the SparseCategoricalCrossentropy function, which expects integer labels, which solved the issue.

In [45]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(32,32,3)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

I fit the model on the training data with 10 epochs.

In [47]:
model.fit(x_train, y_train, epochs = 10)

Epoch 1/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.8768 - accuracy: 0.3112
Epoch 2/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.7101 - accuracy: 0.3831
Epoch 3/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.6515 - accuracy: 0.4046
Epoch 4/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.6060 - accuracy: 0.4198
Epoch 5/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.5613 - accuracy: 0.4392
Epoch 6/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.5315 - accuracy: 0.4473
Epoch 7/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.5009 - accuracy: 0.4594
Epoch 8/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.4752 - accuracy: 0.4682
Epoch 9/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.4590 - accuracy: 0.4727
Epoch 10/10
1563/1563 [==============================] - 8s 5ms/step - loss: 1.438

Evaluate the model with the test data:

In [48]:
test_loss, test_acc = model.evaluate(x_test, y_test, verbose = 2)
print("Test loss: ", test_loss)
print("Test accuracy: ", test_acc)

313/313 - 1s - loss: 1.4872 - accuracy: 0.4688 - 555ms/epoch - 2ms/step
Test loss:  1.487177848815918
Test accuracy:  0.46880000829696655


Baseline model loss and accuracy on test set:
- Loss 1.487
- Accuracy 0.469

Fine-tuned model 1 hyperparameters:
- Hidden layers: 5
- Neurons per hidden layer: 128
- Activation: ReLU
- Optimizer: Adam
- Loss function: SpareCategoricalCrossentropy
- Batch size: 32
- Epochs: 15

In [58]:
finetuned_model1 = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(32,32,3)),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(128, activation='relu'),
  tf.keras.layers.Dense(10)
])

finetuned_model1.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])
finetuned_model1.fit(x_train, y_train, batch_size = 32, epochs = 15)

Epoch 1/15
1563/1563 [==============================] - 9s 5ms/step - loss: 1.8840 - accuracy: 0.3107
Epoch 2/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.7167 - accuracy: 0.3810
Epoch 3/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.6484 - accuracy: 0.4060
Epoch 4/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.5941 - accuracy: 0.4232
Epoch 5/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.5501 - accuracy: 0.4408
Epoch 6/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.5159 - accuracy: 0.4561
Epoch 7/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.4869 - accuracy: 0.4673
Epoch 8/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.4652 - accuracy: 0.4738
Epoch 9/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.4449 - accuracy: 0.4811
Epoch 10/15
1563/1563 [==============================] - 8s 5ms/step - loss: 1.431

In [59]:
ftmodel1_test_loss, ftmodel1_test_acc = finetuned_model1.evaluate(x_test, y_test, verbose = 2)

313/313 - 1s - loss: 1.4555 - accuracy: 0.4821 - 675ms/epoch - 2ms/step


Fine-tuned model 1 loss and accuracy on test set:
- Loss 1.4555
- Accuracy 0.4821